### ISSUE TITLE
template: Abort uninstallation when multiple matches found

### ISSUE THREAD

**Issue by FranklinYu** [chocolatey/choco#1695]
Current template has this:

https://github.com/chocolatey/choco/blob/ea4b155ad4f6911c003fcf30792f85abca480f05/src/chocolatey/infrastructure.app/templates/ChocolateyUninstallTemplate.cs#L92-L97

However, in this case multiple matches are found, and none of them are actually uninstalled, so IMO the installation actually *failed* and should abort. Please add `throw` or something similar to abort uninstallation in this branch.

**Comment by ferventcoder** [comment]
So let me ask, is it appropriate that someone can not uninstall the "package" if there is a bug in the uninstall script that someone wrote? If that package is a dependency of other packages. What if the person uses `--force`? 

Please layout the scenarios of current behavior and what you expect from an end user perspective and then we can discuss which one feels more appropriate.

**Comment by FranklinYu** [comment]
> …is it appropriate that someone can not uninstall the "package" if there is a bug in the uninstall script that someone wrote?

I don’t think we can prevent this. Bugs can occur anywhere; package maintainer may throw exceptions outside of the if-clause.

Concerning the option `--force`, what exactly does it mean? [The documentation](https://chocolatey.org/docs/commands-install) doesn’t provide details:

> Force - force the behavior. Do not use force during normal operation - it subverts some of the smart behavior for commands.

I assume you meant `chocolateyUninstall.ps1` by “uninstall script”. Please correct me if I’m wrong.

**Comment by pauby** [comment]
I'm coming to this very late after looking to issues tagged as 'Waiting On User'.

If I understand correctly, if multiple software matches then the package is uninstalled but the software is not?

If that is the case then that would be my expected behavour. Chocolatey is a package manager and manages packages. The packages manage the software. Chocolatey does not manage the software.

Given that, then Chocolatey should uininstall the package, report the warning that the package may have problems and to be safe it is not going to run the uninstall script.

@gep13 do you have any thoughts on this?

**Comment by TheCakeIsNaOH** [comment]
> If I understand correctly, if multiple software matches then the package is uninstalled but the software is not?

Yes, I believe so.

> Given that, then Chocolatey should uininstall the package, report the warning that the package may have problems and to be safe it is not going to run the uninstall script.

This issue is about code within `chocolateyUninstall.ps1` scripts, so the uninstall script would be already running. 

I think the current behavior is best, as the key(s) passed into this if statement were gathered by `Get-UninstallRegistryKey`. If more than one software entry is returned, it means the package maintainer needs to tweak the `SoftwareName` passed into `Get-UninstallRegistryKey`, or do further filtering of the returned keys. 

In the very rare case in which a single installer creates multiple registry uninstall entries that need to be uninstalled separately, then the maintainer should explicitly change to code to handle it. It is rare enough that I don't think the template should mention it. 

In the case where multiple versions of the software (not the package) can be installed side by side, then again, that is something that the package maintainer should add code to handle explicitly.

https://github.com/chocolatey/choco/blob/ea4b155ad4f6911c003fcf30792f85abca480f05/src/chocolatey/infrastructure.app/templates/ChocolateyUninstallTemplate.cs#L59-L64

**Comment by pauby** [comment]
Thanks for raising this issue. After reviewing it, it's not something we expect to prioritise in the foreseeable future. We’re closing it to keep our backlog focused, but we’re open to revisiting it in the future if priorities shift or new interest arises. You are welcome to comment here or open a new issue if you'd like to discuss it further down the line.
